File: NearlySorted/NearlySorted.java
/*
https://www.geeksforgeeks.org/problems/nearly-sorted-1587115620/1
*/

class Solution {
    public void nearlySorted(int[] arr, int k) {
        int lengthOfArr=arr.length;
        if(lengthOfArr==1) return;
        PriorityQueue<Integer>minHeap=new PriorityQueue<>();
        int initialIndex=0, minWindowSize=Math.min(k+1, lengthOfArr);
        for(; initialIndex<minWindowSize; ++initialIndex){
            minHeap.add(arr[initialIndex]);
        }
        int indexForTraversing=0;
        while(initialIndex<lengthOfArr){
            arr[indexForTraversing++]=minHeap.poll();
            minHeap.add(arr[initialIndex++]);
        }
        while(!minHeap.isEmpty()){
            arr[indexForTraversing++]=minHeap.poll();
        }
    }
}

File: NearlySorted/NearlySorted.cpp
/*
https://www.geeksforgeeks.org/problems/nearly-sorted-1587115620/1
*/

class Solution {
  public:
    void nearlySorted(vector<int>& arr, int k) {
        int sz=arr.size();
        if(sz==1) return;
        priority_queue<int, vector<int>, greater<>>pq;
        int idx=0, minWindowSize=min(k+1, sz);
        for(; idx<minWindowSize; ++idx){
            pq.push(arr[idx]);
        }
        int idx2=0;
        while(idx<sz){
            arr[idx2++]=pq.top();
            pq.pop();
            pq.push(arr[idx++]);
        }
        while(!pq.empty()){
            arr[idx2++]=pq.top();
            pq.pop();
        }
    }
};